homeassistant/components/mqtt/valve.py

```diff
@@ -0,0 +1,386 @@
+            if config.get(CONF_PAYLOAD_CLOSE, DEFAULT_PAYLOAD_CLOSE) is not None:
+                supported_features |= ValveEntityFeature.CLOSE
+
+        if config[CONF_REPORTS_POSITION] and has_command_topic is not None:
+            supported_features |= ValveEntityFeature.SET_POSITION
+        if config.get(CONF_PAYLOAD_STOP) is not None:
+            supported_features |= ValveEntityFeature.STOP
+
+        self._attr_supported_features = supported_features
+
+    @callback
+    def _update_state(self, state: str) -> None:
+        """Update the valve state based on static payload."""
+        self._attr_is_closed = state == STATE_CLOSED
+        self._attr_is_opening = state == STATE_OPENING
+        self._attr_is_closing = state == STATE_CLOSING
+
+    def _prepare_subscribe_topics(self) -> None:
+        """(Re)Subscribe to topics."""
+        topics = {}
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        @write_state_on_attr_change(
+            self,
+            {
+                "_attr_current_valve_position",
+                "_attr_is_closed",
+                "_attr_is_closing",
+                "_attr_is_opening",
+            },
+        )
+        def state_message_received(msg: ReceiveMessage) -> None:
+            """Handle new MQTT state messages."""
+            payload_dict: Any = None
+            position_payload: Any = None
+            state_payload: Any = None
+            payload = self._value_template(msg.payload)
+
+            if not payload:
+                _LOGGER.debug("Ignoring empty state message from '%s'", msg.topic)
+                return
+
+            with suppress(*JSON_DECODE_EXCEPTIONS):
+                payload_dict = json_loads(payload)
+            if isinstance(payload_dict, dict) and "position" in payload_dict:
+                position_payload = payload_dict["position"]
+            if isinstance(payload_dict, dict) and "state" in payload_dict:
+                state_payload = payload_dict["state"]
+            state_payload = payload if state_payload is None else state_payload
+            position_payload = payload if position_payload is None else position_payload
+
+            state: str | None = None
+            reports_position: bool = self._config[CONF_REPORTS_POSITION]
+            if state_payload == self._config[CONF_STATE_OPENING]:
+                state = STATE_OPENING
+            elif state_payload == self._config[CONF_STATE_CLOSING]:
+                state = STATE_CLOSING
+            elif not reports_position and state_payload == self._config.get(
+                CONF_STATE_OPEN, STATE_OPEN
+            ):
+                state = STATE_OPEN
+            elif not reports_position and state_payload == self._config.get(
+                CONF_STATE_CLOSED, STATE_CLOSED
+            ):
+                state = STATE_CLOSED
+            if reports_position and (
+                state is None or position_payload != state_payload
+            ):
```

Comment:
Maybe try splitting the logic to one function which is called if `report_position` is `True` and another if it isn't?